pkg/controller/modeler/modeler.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package modeler implements periodic statistical calculations.
+package modeler
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/gonum/matrix/mat64"
+	"github.com/google/exposure-notifications-verification-server/pkg/config"
+	"github.com/google/exposure-notifications-verification-server/pkg/database"
+	"github.com/google/exposure-notifications-verification-server/pkg/digest"
+	"github.com/google/exposure-notifications-verification-server/pkg/render"
+	"github.com/hashicorp/go-multierror"
+
+	"github.com/google/exposure-notifications-server/pkg/logging"
+
+	"github.com/sethvargo/go-limiter"
+	"go.uber.org/zap"
+)
+
+// Controller is a controller for the modeler service.
+type Controller struct {
+	config  *config.Modeler
+	db      *database.Database
+	h       *render.Renderer
+	limiter limiter.Store
+	logger  *zap.SugaredLogger
+}
+
+// New creates a new modeler controller.
+func New(ctx context.Context, config *config.Modeler, db *database.Database, limiter limiter.Store, h *render.Renderer) *Controller {
+	logger := logging.FromContext(ctx).Named("modeler")
+
+	return &Controller{
+		config:  config,
+		db:      db,
+		h:       h,
+		limiter: limiter,
+		logger:  logger,
+	}
+}
+
+// HandleModel accepts an HTTP trigger and re-generates the models.
+func (c *Controller) HandleModel() http.Handler {
+	logger := c.logger.Named("HandleModel")
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+
+		if err := c.rebuildModels(ctx); err != nil {
+			logger.Errorw("failed to build models", "error", err)
+			c.h.JSON500(w, err)
+			return
+		}
+
+		c.h.RenderJSON(w, http.StatusOK, nil)
+	})
+}
+
+// rebuildModels iterates over all models with abuse prevention enabled,
+// calculates the new limits, and updates the new limits.
+func (c *Controller) rebuildModels(ctx context.Context) error {
+	logger := c.logger.Named("rebuildModels")
+	db := c.db.RawDB()
+
+	// Get all realm IDs in a single operation so we can iterate realm-by-realm to
+	// avoid a full table lock during stats calculation.
+	var ids []uint64
+	if err := db.
```

Comment:
don't love that this query is here and not in the DB package.

c.db.AbuseProtectionEnabledRealms() - or a better name

Reply:
Moved to db